src/plugins/RepairMan.cpp

```diff
@@ -2,15 +2,43 @@
 //
 // Copyright (c) 2017-2020 Alexander Kurbatov
 
+#include "Historican.h"
+#include "Hub.h"
 #include "RepairMan.h"
 #include "core/API.h"
 #include "core/Helpers.h"
 
+#include <sc2api/sc2_unit_filters.h>
+
+namespace {
+Historican gHistory("repairman");
+
+}  // namespace
+
 void RepairMan::OnStep(Builder*) {
-    if (gAPI->observer().GetCurrentRace() != sc2::Race::Terran)
+    if (gHub->GetCurrentRace() != sc2::Race::Terran || m_damaged_buildings.empty())
         return;
 
-    // FIXME (alkuratov): Put buildings repair code here.
+    m_damaged_buildings.remove_if( [](const sc2::Unit* unit_) {
+            return unit_->health == unit_->health_max || !unit_->is_alive;
+        });
+
+    auto repairers = gAPI->observer().GetUnits(IsRepairer());
+    if (repairers().empty() && !m_damaged_buildings.empty()) {
```

Comment:
So it is like one repairer at a time? Shouldn't it be larger count, e.g. one repairer per unit?

Reply:
Also, from performance point of view, it'll be better to bail out on `m_damaged_buildings.empty()` and avoid that heavy GetUnits call.

Reply:
So, when I tested, an enemy army came and damaged like ~5 buildings near ramp and 5 separate workers come to repair the buildings, and they all get killed and then another 5 come and get killed, etc, which quickly killed all my workers. So keeping it to 1 was an attempt to reduce continual worker-deaths.

That's *after* the initial repair attempt though. In OnDamaged, each damaged building gets a repairer, but if it dies, another doesn't get sent unless there are no repairers currently, at which point it will send a repairer to the building that was damaged first.

I think 3 repairers minimum is a better balance, but wanted to start at 1 at least (because it was easier to code). I don't want to put in code that creates a worse behavior than not repairing at all, and having a continual line of workers going to their death seems to be *worse* than no repairing (as it takes a while for enemy to kill buildings, and lets us build more military in the meantime).

For *advanced* repair behavior, it needs to take into account where enemies are, but that need to be done in Marine Micro, or elsewhere first, and is not available yet. I wanted to do a 'minimum' useful Repair Behavior first. After Secure/Defend Expansion Feature and Marine-Micro feature, the defending army should respond to building damage and protect the repairing workers. After that, we could increase the number of repairers.

Another option (for now) is to send the Repairer away if it gets to <50% of its own health.

Reply:
3 repairers looks good to me. 
Agree on <50% health.

Reply:
Ran into a funny bug. Forgot that OnDamaged is called for Enemy units too. So when I attacked an Enemy Terran Turret, it got into m_damaged_buildings, and tried to command self SCVs to Repair, but the command would fail, so SCV would stay on GATHER_MINERALS, but still be in m_busy_workers, and there would be no Repairers, so it would try another SCV, until all other SCV's were in m_busy_workers as REPAIRING.

Reply:
Bug was actually when Minerals = 0, REPAIR command fails and GATHER_MINERALS never stops. Fixed by checking minerals before assigning REPAIR, though this doesn't prevent a sudden construction cost from dropping minerals to 0 on next frame. Fixed that by first command STOP, then queueing REPAIR, so even if REPAIR fails, worker is stopped from GATHER_MINERALS and will then call OnIdle properly.

Reply:
Adding the "<50%" took more additions than I thought. Had to add filter functions to GetClosest methods to prevent trying to grab the same damaged worker that I was sending away.